### PR TITLE
Refactor Test Driver in preparation for `flutter test` integration tests

### PR DESCRIPTION
**Pull request by DanTup**
I've written some simple expression evaluation integration tests for `flutter test` due to #23409 but it required splitting FlutterTestDriver up to be able to share the debugger code without needing to share the daemon protocol (`flutter test` talks its own protocol).

To simplify the test PR, this is a PR just for the refactor. It splits the `FlutterTestDriver` class into a base (that has functionality common to `flutter run` and `flutter test`) and `FlutterRunTestDriver` that has the `flutter run`(/`attach`)-specific bits.

It also renames `TestProject` - the base type for projects used in tests - just to `Project` because the `flutter test` tests have a `TestProject` that is a project with tests.

(@aam @devoncarew)

(Edit: It also extracts the expression evaluation functions further in the test, since they're reusable with the `flutter test` tests)

### CODE REVIEW

**Review by devoncarew — COMMENTED** [review]
This change looks good, but I may not be the best person to review anymore. Likely @aam (also cc'd), and @gspencergoog and @tvolkert could find other reviewers.

File: packages/flutter_tools/test/integration/test_driver.dart [hunk 1]
```diff
@@ -17,12 +17,12 @@ import 'package:web_socket_channel/io.dart';
 import '../src/common.dart';
 
 // Set this to true for debugging to get JSON written to stdout.
-const bool _printJsonAndStderr = false;
+const bool _printJsonAndStderr = true;
```
[the remark this answers is not in the document]
Comment (DanTup) [reply packages/flutter_tools/test/integration/test_driver.dart]: Er no.. I undid it somewhere, but I guess it wasn't here! I'll revert before merging, thanks!

**Comment by DanTup** [comment]
If it's useful, these are the new tests that these changes were made to support:

https://github.com/DanTup/flutter/commit/a4ba9432ace0f5fbc43de754aa584777c1a6f781

(I haven't opened a PR yet as it'd also includes these changes and I wanted to keep the reviews smaller)

**Comment by DanTup** [comment]
(this PR conflicts badly with the vm_service_lib changes I just landed so will need some work!)

**Comment by DanTup** [comment]
(rebased on top of the vm_service_lib changes, good to review now)

**Review by tvolkert — APPROVED** [review]
LGTM

**Pull request merged**